Region/Region.cc

```diff
@@ -364,16 +364,55 @@ casacore::IPosition Region::xyShape() {
 // ***********************************
 // Region data
 
-bool Region::getData(std::vector<float>& data, casacore::SubLattice<float>& sublattice) {
+bool Region::getData(std::vector<float>& data, casacore::SubLattice<float>& sublattice, int checkPerChannels) {
     // fill data vector using region SubLattice
     bool dataOK(false);
     casacore::IPosition sublattShape = sublattice.shape();
     data.resize(sublattShape.product());
-    casacore::Array<float> tmp(sublattShape, data.data(), casacore::StorageInitPolicy::SHARE);
-    try {
-        sublattice.doGetSlice(tmp, casacore::Slicer(casacore::IPosition(sublattShape.size(), 0), sublattShape));
-        dataOK = true;
-    } catch (casacore::AipsError& err) {
+    if (checkPerChannels > 0 && sublattShape.size() > 2) { // stoppable spectral profile process
+        try {
+            casacore::IPosition start(sublattShape.size(), 0);
+            casacore::IPosition count(sublattShape);
+            int profileAxis; // profile axis index
+            size_t profileSize; // profile vector size
+            // get profile axis index and its vector size
+            for (int i=0; i<sublattShape.size(); ++i) {
+                if (count(i)>1) {
+                    profileAxis = i;
+                    profileSize = count(i);
+                }
+            }
+            size_t begin = 0; // the begin index of profile vector in each copy
+            size_t upperBound = (profileSize%checkPerChannels == 0 ? profileSize/checkPerChannels : profileSize/checkPerChannels + 1);
+            // get regional parameters setting as this moment
+            std::vector<CARTA::Point> tmp_ctrlpoints = m_ctrlpoints;
```

Comment:
do we need a mutex around this? What if one thread is busy updating a region, while the other thread is calculating a spectral profile?

Reply:
There is a mutex start from https://github.com/CARTAvis/carta-backend/blob/master/Frame.cc#L738. So I think this block is also been locked.

Reply:
Doesn't that mean that there's no way for `m_ctrlpoints` to ever be changed during `Region::getData`?

Reply:
just reading through the code, that lock is the `latticeMutex` which the frame does not attempt to lock when updating region parameters. So it's still possible to update `m_ctrlpoints` while `getData` is running, which makes sense. However, my concern is that there's a very small chance that you could be copying `m_ctrlpoints` to `tmp_ctrlpoints` _while_ `m_ctrlpoints` is being updated by `setPoints`

Unrelated, note that changing region rotation will also invalidate the current region calculations

Reply:
I see. I am thinking a good way to prevent this case happens.

Reply:
it's a very unlikely situation, but I guess it _could_ happen?

Reply:
This could happen, as Session::onSetRegion sends the spatial profile before initiating the spectral profile.  The calls for data in onSetRegion could include its region points/rotation and used to check for change (not only for profiles, but also stats and histograms).  Or perhaps SubLattice::positionInParent would tell us these xy coordinates?